### DISCUSSION TITLE
mv/cp in background

### DISCUSSION THREAD

**Discussion by Docbroke** [leo-arch/clifm#93] [General]
Can we have mv and cp run in background, and free clifm to do other things? This is particularly needed when copying multiple large files to different partitions or remote locations.
This may be different from what we expect when running mv/cp in terminal, but clifm is also a file-manager, so it is little counterintuitive to wait for mv/cp to finish before using clifm for other tasks. Sure we can start another instance of clifm but we still need to open all needed tabs and dirs.

**Comment by leo-arch** [comment]
Why not to use the `x` command? It will run a new instance of CliFM in the same path you're currently in. There you can fire up cp/mv and keep working as usual in the first instance.

One drawback nonetheless: this won't work for non-graphical environments.

**Reply by Docbroke** [reply]
another drawback,  `x`  starts new instance with only one tab. When I do copy/paste, I usually have 2-3 tabs open, I select from 1 tab and `cp/mv sel` in other tabs.

**Reply by leo-arch** [reply]
Are you talking about terminal tabs or clifm workspaces?

**Reply by Docbroke** [reply]
clifm workspaces

**Reply by Docbroke** [reply]
I think I can select files in one instance of clifm and use other instances to copy/move, that way only one instance of clifm will be blocked. Instead of using 2-3 workspaces in the same instance.

**Reply by leo-arch** [reply]
I see. I think it's a good idea: it would improve the workflow. But I'm not sure about the implementation: having cp/mv in the background won't allow you to follow the process because STDOUT is silenced. This is also counterintuitive.

The ideal solution would be to have some kind of extra window showing the cp/mv thing progress, while liberating the first instance to keep doing whatever you're doing.

**Reply by leo-arch** [reply]
> I think I can select files in one instance of clifm and use other instances to copy/move

That would be an acceptable approach, but only for big files. For small ones workspaces work just fine.

**Reply by Docbroke** [reply]
yes indeed. This sounds good enough. For small files, workspaces, for large files and remote destinations, new instance.

**Reply by leo-arch** [reply]
What could be done, though I'm just thinking loud, is to launch cp/mv in a second terminal window. But I guess this would be a bit odd and annoying when copying small files: a terminal window would be launched to immediately dissapear.

**Reply by Docbroke** [reply]
I agree that backgrounding cp/mv is not very good idea, unless we are able to add some indicator of copy/mv progress and mechanism to stop it if needed, which probably is unwanted complication.

Popping terminals will surely annoy users. I remember when I used `midnight commander` there was a command to background copy/move progress. So if copying is taking time you can use keyboard shortcut to background that and keep using `mc`. However that probably used curses.

**Reply by leo-arch** [reply]
Well, I think you can do something similar, but needs to be tested: if cp/mv is taking too much, you can press C-x to launch a new instance. However, it needs to be tested: right now I cannot guarantee that this keybinding will be available while running a shell command. Maybe.

EDIT: No, it's not. For this to be done you need multi-threading.
